pdns/recursordist/lwres.cc

```diff
@@ -382,11 +483,106 @@ static void addPadding(const DNSPacketWriter& pw, size_t bufsize, DNSPacketWrite
+{
+  auto lock = s_cookiestore.lock();
+  auto found = lock->find(address);
+
+  if (found == lock->end()) {
+    // We received cookie (we might have sent one out) but the server is not in the table?
+    // This is a case of cannot happen, unless rec_control clear-cookies was called
+    VLOG(log, "Cookie from " << address.toString() << " not found back in table" << endl);
+    lwr.d_rcode = RCode::FormErr;
+    lwr.d_validpacket = false;
+    return {true, LWResult::Result::Success}; // success - oddly enough
+  }
+
+  // We have stored cookie info, scan for COOKIE option in EDNS
+  if (const auto opt = edo.getFirstOption(EDNSOptionCode::COOKIE); opt != edo.d_options.end()) {
+    if (EDNSCookiesOpt received; received.makeFromString(opt->second)) {
+      cookieFoundInReply = true;
+      VLOG(log, "Received cookie info back from " << address.toString() << ": " << received.toDisplayString() << endl);
+      if (received.getClient() == cookieSentOut->getClient()) {
+        VLOG(log, "Client cookie from " << address.toString() << " matched! Storing with localAddress " << localip.toString() << endl);
+        ++t_Counters.at(rec::Counter::cookieMatched);
+        found->d_localaddress = localip;
+        found->d_cookie = received;
+        if (found->getSupport() == CookieEntry::Support::Probing) {
+          ++t_Counters.at(rec::Counter::cookieProbeSupported);
+        }
+        found->setSupport(CookieEntry::Support::Supported, now.tv_sec);
+        // check extended error code
+        uint16_t ercode = edo.getCombinedERCode(lwr.d_rcode);
+        if (ercode == ERCode::BADCOOKIE) {
+          lwr.d_validpacket = true;
+          ++t_Counters.at(rec::Counter::cookieRetry);
+          VLOG(log, "Server " << localip.toString() << " returned BADCOOKIE " << endl);
+          return {true, LWResult::Result::BadCookie}; // We did update the entry, retry should succeed
+        }
+      }
+      else {
+        if (!doTCP) {
+          // Server responded with a wrong client cookie, fall back to TCP, RFC 7873 5.3
+          VLOG(log, "Server " << localip.toString() << " responded with wrong client cookie, fall back to TCP" << endl);
+          lwr.d_validpacket = true;
+          ++t_Counters.at(rec::Counter::cookieMismatchedOverUDP);
+          return {true, LWResult::Result::Spoofed};
+        }
+        // mismatched cookie when already doing TCP, ignore that
+        VLOG(log, "Server " << localip.toString() << " responded with wrong client cookie over TCP, ignoring that" << endl);
+        ++t_Counters.at(rec::Counter::cookieMismatchedOverTCP);
+      }
+    }
+    else {
+      VLOG(log, "Malformed cookie in reply from " << address.toString() << ", dropping as if was a timeout" << endl);
+      // Do something special if we get malformed repeatedly? And/or consider current status?
```

Comment:
yes, perhaps add a counter and timer and remove SUPPORTED status after 3 failures within a minute or so?